src/app/model-poller.js

```diff
@@ -99,3 +60,75 @@ export default async function connectAndListModels(
     console.error("Something went wrong: ", error);
   }
 }
+
+/**
+
+  @param {Object} controllerData The data to use to connect to the controller.
+    In the format [
+      wsControllerURL - The fully qualified controller url wss://ip:port/api
+      credentials - An object with the keys {user, password}
+      bakery - An instance of the bakery to use if necessary
+      identityProviderAvailable - If an identity provider is to be used. If so
+        a bakery must be provided.
+    ]
+  @param {Boolean} isJuju
+  @param {Object} reduxStore
+*/
+export async function connectAndPollController(
+  controllerData,
+  isJuju,
+  reduxStore
+) {
+  let conn, error, juju, intervalId;
+  try {
+    ({ conn, error, juju, intervalId } = await loginWithBakery(
+      ...controllerData
+    ));
+    if (error) {
+      reduxStore.dispatch(storeLoginError(error));
+      return;
+    }
+  } catch (e) {
+    return console.log("unable to log into controller", e, controllerData);
+  }
+
+  // XXX Now that we can register multiple controllers this needs
+  // to be sent per controller.
+  if (process.env.NODE_ENV === "production") {
+    Sentry.setTag("jujuVersion", conn?.info?.serverVersion);
+  }
+
+  reduxStore.dispatch(updateControllerConnection(controllerData[0], conn));
+  reduxStore.dispatch(updateJujuAPIInstance(controllerData[0], juju));
+  reduxStore.dispatch(updatePingerIntervalId(controllerData[0], intervalId));
+
+  fetchControllerList(controllerData[0], conn, controllerData[4], reduxStore);
+  // XXX the isJuju Check needs to be done on a per-controller basis
+  if (!isJuju) {
+    // This call will be a noop if the user isn't an administrator
+    // on the JIMM controller we're connected to.
+    disableControllerUUIDMasking(conn);
+  }
+
+  do {
+    const models = await conn.facades.modelManager.listModels({
```

Comment:
All of this was simply moving the code to a separate function to make it easier to reason about.

Changes start from here (114) to 126